File: codeforces/1800~1899/1860/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  string s;
  cin >> s;

  int n = s.length();

  constexpr int INF = 1e9;
  vector<vector<int>> dp(n+1, vector<int>(n*n, INF));
  dp[0][0] = 0;
  for (int i = 0; i < n; ++i) {
    vector<vector<int>> next_dp(n+1, vector<int>(n*n, INF));

    for (int j = 0; j <= i; ++j) {
      for (int c01 = 0; c01+j < n*n; ++c01) {
        next_dp[j+1][c01] = min(next_dp[j+1][c01], dp[j][c01] + (s[i] != '0'));
        next_dp[j][c01+j] = min(next_dp[j][c01+j], dp[j][c01] + (s[i] != '1'));
      }
    }

    dp = move(next_dp);
  }

  int c0 = count(s.begin(), s.end(), '0');
  cout << dp[c0][c0 * (n-c0) / 2] / 2<< endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}

File: codeforces/1800~1899/1860/d.rs
use std::io;
use std::cmp::min;

const INF: i32 = 1e9 as i32;

fn read_str() -> String {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn solve() {
    let s = read_str();
    let n = s.len();
    let mut dp: Vec<Vec<i32>> = vec![vec![INF; n*n]; n+1];
    dp[0][0] = 0;
    for i in 0..n {
        let mut next_dp: Vec<Vec<i32>> = vec![vec![INF; n*n]; n+1];
        let ch = s.chars().nth(i).unwrap();
        for j in 0..=(i as usize) {
            let mut c01: usize = 0;
            while c01+j < n*n {
                next_dp[j+1][c01] = min(next_dp[j+1][c01], dp[j][c01] + ((ch != '0') as i32));
                next_dp[j][c01+j] = min(next_dp[j][c01+j], dp[j][c01] + ((ch != '1') as i32));
                c01 += 1;
            }
        }

        dp = next_dp;
    }

    let c0 = s.chars().filter(|ch| *ch == '0').count() as usize;
    println!("{}", dp[c0][c0 * (n-c0) / 2] / 2);
}

fn main() {
    solve();
}
